README.md
<p align="center">
<a href="https://github.com/7Cav/SandboxAndTrainingMissionGenerator/releases/latest"><img src="https://img.shields.io/github/release/7Cav/SandboxAndTrainingMissionGenerator.svg?style=for-the-badge&label=Release%20Build" alt="Release Build Version"></a>
<a href="https://github.com/7Cav/SandboxAndTrainingMissionGenerator/releases/latest"><img src="https://img.shields.io/github/downloads/7cav/SandboxAndTrainingMissionGenerator/total.svg?style=for-the-badge&label=Downloads" alt="Downloads"></a>
<a href="https://travis-ci.org/7Cav/SandboxAndTrainingMissionGenerator"><img src="https://img.shields.io/travis/7Cav/SandboxAndTrainingMissionGenerator.svg?style=for-the-badge&logo=Travis-CI" alt="Build"></a>
</p>

This is a mission generator script built to quickly and reliable build sandbox and training missions for the 7th Cavalry Gaming.

# Requirements
* [Python3](https://www.python.org)
* [armake](https://github.com/KoffeinFlummi/armake)

## Install
**Windows:** 
- [Download and install python3](https://www.python.org)
- [Download and install armake](https://github.com/KoffeinFlummi/armake/releases/latest)
  
**Linux:**
```
$ sudo apt-get install python3
```
```
$ sudo add-apt-repository ppa:koffeinflummi/armake
$ sudo apt-get update
$ sudo apt-get install armake
```

# How to run
```
usage: build [-h] -p PACKAGE [-pv PACKAGEVERSION] -v VERSIONTAG [-y] [--color]
             [-m MISSION] [-s SETTING] [--version]
             {sandbox,training}

This script generates missions.

positional arguments:
  {sandbox,training}    This defines what kind of generation the script should
                        commit.

optional arguments:
  -h, --help            show this help message and exit
  -p PACKAGE, --package PACKAGE
                        This defines what script package to install.
  -pv PACKAGEVERSION, --packageVersion PACKAGEVERSION
                        This define a version number of a given package to be
                        used in the script.
  -v VERSIONTAG, --versionTag VERSIONTAG
                        This define what version name you whant the file to
                        have.
  -y, --fastbuild       Will instantly run untill done.
  --color               Enable colors in the script.
  -m MISSION, --mission MISSION
                        Define a specific mission from path
  -s SETTING, --setting SETTING
                        Define a custom setup.json file
  --version             show program's version number and exit

This build script generates sandboxes or training mapes bases on avalible templates.
The tool should be cross platform and can be used for other packages as well.
```
<!--- (Soon) Modfify the `properties.ini` if needed.-->
- Modify the `setup.json` if needed.
- Modify the Templates if needed. (See below for requirements.) 
- Run the script<br />
  Windows: `py build.py sandbox -v 1.0 -p cScripts.zip` or
           `py build.py training -v 1.0 -p cScripts.zip`<br />
  Linux: `python3 build.py sandbox -v 1.0 -p cScripts.zip` or
         ` python3 build.py training -v 1.0 -p cScripts.zip`

## Setting up a sandbox template
- __Mission file most be unbinirized__.
- Script will primarily use the Generic template.<br />
  A custom template can be used to create one the folowing name is required:<br />
  `Template_Altis.Altis` or `Template_MyIsland.MyIsland`<br />
  The island need to be placed in the `./Template/sandbox/` directory.
- To allow the script to set the Respawn use the following coordinates:<br />
  `position[]={20.200001,25.200001,20.200001};` 
- Unit placement is recommended to be set in the lower left corner on short grid `00 00`.

## Setting up a training mission
- __Mission file most be unbinirized__.
- Set the `briefingName` name to `Training Template Mission`
- Training missions name need to be in the following format `[Company]_CO_Trg_[My_Training_Mission_Name-Map]_DEVBUILD.[Island_Name]`<br />Example: `7cav_Charlie_CO_Trg_Map_DEVBUILD.Stratis` or `7cav_Charlie_CO_Trg_Ambush_DEVBUILD.Altis` 
- Training missions need to placed in `./template/training/`.
- Additional training mission scripts images or other material need to be placed in the mission root folder. 
- Adjust or add a `setup.json`<br />__NOTE!__ Do not add your own `init.sqf` of `description.ext` they will be overwritten. instead add changes or adjustmetns instead to the `add : []` array of the json file.

## Custom build
The custom mission build is preformed when you use the `training` and `-mission [mission]` parameters.
To run this you need to be placed the mission in; `./template/training/`. The build follows the training mission build system. But instead of building all missions it only build for one. Exsample:<br />
`$ build.py training -v 1.0 -p cScripts.zip -m My_Mission_Name.Island`

## String hooks
String hooks refere to names that the script look for in order to replace a Line. If it can't fine the string it will skip the change.

### mission.sqm
- **briefingName:** `$briefingName`, `Zeus Sandbox Template Mission`, `Training Template Mission` (string)
- **overviewText:** `$overviewText`, `OverviewText Template Text` (string)
- **Spawn point move pos:** `position[]={20.200001,25.200001,20.200001};`

### description.ext
<!--
- **author:** `$author`, `1SG Tully.B`
- **onLoadName** `MyMissionName`
- **onLoadMission** `7th Cavalry - S3 1BN Battle Staff Operation`
- **onLoadIntro** `S3 1BN Battle Staff Operation`
- **loadScreen** `Data\MissionLogo.paa`
- **overviewPicture** `Data\MissionLogo.paa`
- **cba_settings_hasSettingsFile** `cba_settings_hasSettingsFile = 1;`
- **disabledAI** `disabledAI              = true;`
-->

### Magicwords
- `$1`: Print version number provided by `--versionTag` parameter.
- `$2`: Print name of the installed package selected by `--packageVersion` parameter.


CHANGELOG.md
# CHANGELOG
## Version 2.2.4
- Fixed issue were generator names returned `none` if a hook was not pressent.
- Added Levonia to generator mission list
- Added ability to generate selected or custom missions
- Added ability to generate sandboxes with custom json
- Updated cScripts to v4.3.9

## Version 2.2.3
- Fixed cba_settings params in description.ext getting changed to empty array when not defined.

## Version 2.2.2
- Updated cScripts to v4.3.8 Hotfixed

## Version 2.2.1
- Added proper naming and versioning to training missions
- cScripts version is now visible in the mission selection
- Updated cScripts to v4.3.8

## Version 2.2.0
- Added Training missions
- Fixed mission controller not working

## Version 2.1.2
- Updated cScripts to v4.3.7
- Updated Generic Template with new classnames

## Version 2.1.1
- Updated cScripts to v4.3.6

## Version 2.1.0
- Updated cScripts to v4.3.5
- Updated unit deployment
- Added new maps

## Version 1.1.0
- Updated cScripts to v4.3.2

## Version 1.0.18
- Updated cScripts to v4.2.15

## Version 1.0.17
- Updated cScripts to v4.2.14

## Version 1.0.16
- Updated cScripts to v4.2.13

## Version 1.0.15
- Updated cScripts to v4.2.11

## Version 1.0.14
- Updated cScripts to v4.2.10

## Version 1.0.13
- Updated cScripts to v4.2.9
- Added new maps

## Version 1.0.12
- Updated cScripts to v4.2.9.RC1

## Version 1.0.11
- Removed the weird `CH_47_F` class referenced in the mission.sqm
- (Hotfixed) Amount of mod items were showing 7 instead of 6

## Version 1.0.10
- Updated cScripts to 4.2.7.RC3 DevBuild

## Version 1.0.9
- Updated cScripts to 4.2.7.RC2 DevBuild

## Version 1.0.8
- Updated cScripts to 4.2.6.RC1 DevBuild
- Change script algorithms to handle CBA_Settings differently adding cScripts settings.
- Fixed cba_settings for cScripts mission type being wrong.
- Fixed spawn point on Malden.

## Version 1.0.6
- Updated cScripts to 4.2.4.DevBuild
- Adjusted cba_settings some.

## Version 1.0.5
- Updated cScripts to 4.2.3.DevBuild
- Removed S3 immortality.

## Version 1.0.4
- Updated cScripts to 4.2.2
- Added mission menu text.

## Version 1.0.3
- Updated cScripts

## Version 1.0.2
- Fixed issue were respawn not being initzialised at start of the mission.
- Updated slot selection with all cav avalible units.
- Named zeus moduels and name synced them to the add to zeus moduel.

## Version 1.0.1
- Missions now have a better spawn position
- Synced the spawn moduel to zeus(edited)


build.py
#!/usr/bin/env python3
import sys, os, shutil, subprocess, time
import argparse, tempfile, zipfile, fileinput, json
startTime = time.time()

__version__ = 3.1

scriptPath = os.path.realpath(__file__)
scriptDir = os.path.dirname(scriptPath)
outputDir = '{}/output'.format(scriptDir)
releaseDir = '{}/release'.format(scriptDir)
templateDir = '{}/template'.format(scriptDir)

# #########################################################################################

parser = argparse.ArgumentParser(
    prog='build',
    formatter_class=argparse.RawDescriptionHelpFormatter,
    description='This script generates missions.',
    epilog='This build script generates sandboxes or training mapes bases on avalible templates.\nThe tool should be cross platform and can be used for other packages as well.'
)

parser.add_argument('buildtype',
    choices=['sandbox', 'training'],
    help='This defines what kind of generation the script should commit.'
)

parser.add_argument('-p', '--package',
    required=True,
    help='This defines what script package to install.'
)
parser.add_argument('-pv', '--packageVersion',
    help='This define a version number of a given package to be used in the script.'
)

parser.add_argument('-v', '--versionTag',
    required=True,
    help='This define what version name you whant the file to have.'
)

parser.add_argument("-y", "--fastbuild",
    help="Will instantly run untill done.",
    action="store_false"
)
parser.add_argument('--color',
    help='Enable colors in the script.',
    action='store_true'
)

parser.add_argument('-m', '--mission',
    default='',
    required=False,
    help='Define a specific mission from path'
)
parser.add_argument('-s', '--setting',
    default='',
    required=False,
    help='Define a custom setup.json file'
)

parser.add_argument('--version', action='version', version='Author: Andreas Broström <andreas.brostrom.ce@gmail.com>\nScript version: {}'.format(__version__))

args = parser.parse_args()



# handle arguments
PACKAGE = args.package
SELECTED_MISSION = args.mission
SELECTED_JSON = args.setting

VERSION = args.versionTag
VERSION_DIR = args.versionTag.replace('.','_')

PACKAGEVER = args.package
if not args.packageVersion == '':
    PACKAGEVER = args.packageVersion

# #########################################################################################

def color_string(string='', color='\033[0m', use_color=False):
    if use_color:
        return '\033[0m{}{}\033[0m'.format(color,string)
    else:
        return string


def build_pbo(temp_folder='', pbo_name='unnamed', use_color=False):
    os.chdir(scriptDir)
    print('Building and compiling {}...'.format(color_string('{}.pbo'.format(pbo_name),'\033[96m',use_color)))
    subprocess.call('armake build -f -p "{}" "output/{}.pbo"'.format(temp_folder,pbo_name), shell=True)


def build_archive(archive_name='unnamed', archive_type='zip', archive_input='', use_color=False):
    print('Archiving all generated mission files...')
    archive_output = '{}/{}'.format(releaseDir,archive_name)
    shutil.make_archive(archive_output, archive_type, archive_input)
    fullarchname = color_string('{}.{}'.format(archive_name,archive_type),'\033[96m',use_color)
    print('{} have been created...'.format(fullarchname))


def check_or_create_folder(dir=''):
    if not os.path.exists(dir):
        os.makedirs(dir)


def get_templates(template_type='', world=''):
    if template_type == 'sandbox':
        if os.path.exists('{0}/sandbox/Template_{1}.{1}'.format(templateDir,world)):
            print('Mission files found for {} using them instead of generic...'.format(world))
            return '{0}/sandbox/Template_{1}.{1}'.format(templateDir,world)
        print('Creating mission file with generic template...'.format(world))
        return '{0}/sandbox/Template_Generic.VR'.format(templateDir)
    elif template_type == 'training':
        return os.listdir('{0}/training/'.format(templateDir))
    else:
        sys.exit('ERROR: No valid template defined')


def fetch_objects(template_path=''):
    os.chdir(template_path)
    content = os.listdir(template_path)

    objectList = []

    folderList = [] # Collect directories
    fileList = []   # Collect files

    for obj in content:
        if os.path.isfile(obj):
            fileList.append(obj)
        elif os.path.isdir(obj):
            folderList.append(obj)
        else:
            sys.exit('\nIssues occured when listing files.')

    objectList.append(folderList)
    objectList.append(fileList)

    os.chdir(scriptDir)

    return objectList


def install_script_package(script_package, temp_path, use_color=False):
    print('Installing script package {}...'.format(color_string(script_package,'\033[96m',use_color)))
    script_package_full_path = '{}/{}'.format(scriptDir,script_package)
    try:
        scriptsArchive = zipfile.ZipFile(script_package_full_path, 'r')
    except:
        sys.exit("Could not locate script package in the root directory...")
    scriptsArchive.extractall(temp_path)
    scriptsArchive.close()


def replace(file, searchExp, replaceExp):
    for line in fileinput.input(file, inplace=1):
        if searchExp in line:
            line = line.replace(searchExp,replaceExp)
        sys.stdout.write(line)


def additions(file, additions=[]):
    with open(file, 'a') as add:
        add.write('\n')
        for line in additions:
            add.write('\n{}'.format(line))
    add.close()


def json_macro_replace(obj):
    if type(obj) == str:
        if '$0' in obj:
            return obj.replace('$0', '{}'.format(VERSION))
        if '$1' in obj:
            return obj.replace('$1', '{}'.format(PACKAGEVER))
    elif type(obj) == list:
        new_list = []
        for line in obj:
            line = line.replace('$0', '{}'.format(VERSION))
            line = line.replace('$1', '{}'.format(PACKAGEVER))
            new_list.append(line)
        return new_list
    else:
        sys.exit('ERROR: JSON macro function can\'t handle input...')
    return obj

def clean_mission_folder(temp_folder=''):
    os.chdir(temp_folder)
    remove_list = ['README.md', 'setup.json', 'Data/MissionLogo.psd']
    for f in remove_list:
        try:
            os.remove(f)
        except:
            pass

def cleanup_output():
    print('Cleaning up output folder...')
    output_list = os.listdir(outputDir)
    os.chdir(outputDir)
    for f in output_list:
        os.remove(f)
    os.chdir(scriptDir)


def setup_missions(temp_folder='', sandbox_json_data={}, count=0, use_color=False):
    os.chdir(temp_folder)

    if args.buildtype == 'sandbox':
        world_spawn_list = []
        for world in sandbox_json_data['worlds']:
            world_spawn_list.append(sandbox_json_data['worlds'][world])

        spawn = world_spawn_list[count]

    # Reset common variable
    changes = ''
    filename = 'mission.sqm'
    file = '{}/{}'.format(temp_folder,filename)
    if filename in sandbox_json_data and os.path.isfile(file):
        print('Applying adjustmetns to {}...'.format(color_string(filename,'\033[96m',use_color)))

        # Spawn
        if args.buildtype == 'sandbox':
            print('Spawn set to {}, {}, {}.'.format(color_string(spawn[0],'\033[92m',use_color), color_string(spawn[1],'\033[92m',use_color), color_string(spawn[2],'\033[92m',use_color)))
            replace(file,
                'position[]={20.200001,25.200001,20.200001};',
                'position[]={{{},{},{}}};'.format(spawn[0],spawn[1],spawn[2]))
        
        for changes in sandbox_json_data[filename]:
            string = sandbox_json_data[filename][changes] 
            string = json_macro_replace(string)

            if 'briefingName' == changes:
                replace(file,
                'briefingName="$briefingName";',
                'briefingName="{}";'.format(string))

                replace(file,
                'briefingName="Zeus Sandbox Template Mission";',
                'briefingName="{}";'.format(string))

                replace(file,
                'briefingName="Training Template Mission";',
                'briefingName="{}";'.format(string))
                continue

            if 'overviewText' == changes:
                replace(file,
                'overviewText="$overviewText";',
                'overviewText="{}";'.format(string))

                replace(file,
                'overviewText="OverviewText Template Text";',
                'overviewText="{}";'.format(string))
                continue

    else:
        '' if os.path.isfile(file) else sys.exit('ERROR: No {} discoverd something is terrible wrong'.format(color_string(filename,'\033[96m',use_color)))
        print('No changes to {} is defined skipping this step...'.format(color_string(filename,'\033[96m',use_color)))


    # Reset common variable
    changes = ''
    filename = 'description.ext'
    file = '{}/{}'.format(temp_folder,filename)
    if os.path.isfile(file):
        print('Applying adjustmetns to {}...'.format(color_string(filename,'\033[96m',use_color)))

        for changes in sandbox_json_data[filename]:
            string = sandbox_json_data[filename][changes]
            string = json_macro_replace(string)

            if changes == 'author':
                replace(file,
                    'dev                 = "$author";',
                    'dev                 = "{}";'.format(string))
                replace(file,
                    'author              = "$author";',
                    'author              = "{}";'.format(string))

                replace(file,
                    'dev                 = "1SG Tully.B";',
                    'dev                 = "{}";'.format(string))
                replace(file,
                    'author              = "1SG Tully.B";',
                    'author              = "{}";'.format(string))
                continue

            if changes == 'onLoadName':
                replace(file,
                    'onLoadName          = "MyMissionName";',
                    'onLoadName          = "{}";'.format(string))
                continue

            if changes == 'onLoadMission':
                replace(file,
                    'onLoadMission       = "7th Cavalry - S3 1BN Battle Staff Operation";',
                    'onLoadMission       = "{}";'.format(string))
                continue

            if changes == 'onLoadIntro':
                replace(file,
                    'onLoadIntro         = "S3 1BN Battle Staff Operation";',
                    'onLoadIntro         = "{}";'.format(string))
                continue

            if changes == 'loadScreen':
                replace(file,
                    'loadScreen          = "Data\MissionLogo.paa";',
                    'loadScreen         = "{}";'.format(string))
                continue

            if changes == 'overviewPicture':
                replace(file,
                    'overviewPicture     = "Data\MissionLogo.paa";',
                    'overviewPicture     = "{}";'.format(string))
                continue

            if changes == 'cba_settings_hasSettingsFile':
                replace(file,
                    'cba_settings_hasSettingsFile = 1;',
                    'cba_settings_hasSettingsFile = {};'.format(string))
                continue

            if changes == 'disabledAI':
                replace(file,
                    'disabledAI              = true;',
                    'disabledAI              = {};'.format(string))
                continue

            if changes == 'forceRotorLibSimulation':
                replace(file,
                    'forceRotorLibSimulation = 1;',
                    'forceRotorLibSimulation = {};'.format(string))
                continue

            if changes == 'respawn':
                replace(file,
                    'respawn                = BASE;',
                    'respawn                = {};'.format(string))
                continue

            if changes == 'respawnDelay':
                replace(file,
                    'respawnDelay           = 4;',
                    'respawnDelay           = {};'.format(string))
                continue

            if changes == 'respawnOnStart':
                replace(file,
                    'respawnOnStart         = -1;',
                    'respawnOnStart         = {};'.format(string))
                continue

            if changes == 'add':
                additions(file, string)
                continue

    else:
        print('No changes to {} is defined skipping this step...'.format(color_string(filename,'\033[96m',use_color)))


    # Reset common variable
    changes = ''
    filename = 'init.sqf'
    file = '{}/{}'.format(temp_folder,filename)
    if os.path.isfile(file):
        print('Applying adjustmetns to {}...'.format(color_string(filename,'\033[96m',use_color)))
       
        for changes in sandbox_json_data[filename]:
            string = sandbox_json_data[filename][changes] 
            string = json_macro_replace(string)
            
            if changes == 'add':
                additions(file, string)
                continue

    else:
        print('No changes to {} is defined skipping this step...'.format(color_string(filename,'\033[96m',use_color)))


    # Reset common variable
    changes = ''
    filename = 'cba_settings.sqf'
    file = '{}/{}'.format(temp_folder,filename)
    if os.path.isfile(file):
        print('Applying adjustmetns to {}...'.format(color_string(filename,'\033[96m',use_color)))
       
        for changes in sandbox_json_data[filename]:
            string = sandbox_json_data[filename][changes] 
            string = json_macro_replace(string)

            if changes == 'add':
                additions(file, string)
                continue

    else:
        print('No changes to {} is defined skipping this step...'.format(color_string(filename,'\033[96m',use_color)))


# #########################################################################################


def main():
    # setup directories if non exist
    check_or_create_folder('release')
    check_or_create_folder('output')
    check_or_create_folder('template/sandbox')
    check_or_create_folder('template/training')

    if args.buildtype == 'sandbox':

        # Exit on non supported params
        sys.exit('Sandbox does not support selected builds using the \'--mission\' parameter exiting...') if not SELECTED_MISSION == '' else ''

        # get json data
        sandbox_json_setup = 'setup.json'
        if SELECTED_JSON:
            sandbox_json_setup = SELECTED_JSON
        print('Fetching data from {}...'.format(color_string(sandbox_json_setup,'\033[96m',args.color)))

        sandbox_json = '{}/{}'.format(scriptDir, sandbox_json_setup)
        with open(sandbox_json) as json_file:  
            sandbox_data = json.load(json_file)
 
        # get mission name
        sandbox_mission_name = sandbox_data['sandboxMissionName']

        # set up world list
        world_list = []
        for world in sandbox_data['worlds']:
            world_list.append(world)

        for count, world in enumerate(world_list):
            temp_path = tempfile.mkdtemp()
            mission_name = '{}_v{}.{}'.format(sandbox_mission_name, VERSION_DIR, world)

            print('Creating sandbox mission on {}... ({}/{})'.format(color_string(world,'\033[92m',args.color), count+1, len(world_list)))

            template_path = get_templates('sandbox', world)
            content_list = fetch_objects(template_path)
            folder_list = content_list[0]
            file_list = content_list[1]

            os.chdir(template_path)

            for obj in folder_list:
                shutil.copytree(obj, '{}/{}'.format(temp_path, obj))
            for obj in file_list:
                shutil.copy2(obj, temp_path)

            # Unzip and install script package
            install_script_package(PACKAGE, temp_path, args.color)

            # Setup mission file
            print('Setting up and adjusting sandbox mission file...')
            setup_missions(temp_path, sandbox_data, count, args.color)

            # Cleaning mission from readme and template files.
            clean_mission_folder(temp_path)
            
            # Building PBO
            build_pbo(temp_path, mission_name, args.color)


    if args.buildtype == 'training':
        all_templates = get_templates('training')
        all_templates.remove('setup_template.json')
        template_dir_name = 'training'
        
        # Exit on non supported params

        sys.exit('Training missions does not support selected json using the \'--setting\' parameter exiting...') if not SELECTED_JSON == '' else ''

        # over write all found templates if mission is defined
        if SELECTED_MISSION:
            template_dir_name = 'custom'
            all_templates = [SELECTED_MISSION]

        sys.exit('No training missions found in "./template/training/"...') if (len(all_templates) == 0) else ''

        for count, world in enumerate(all_templates):
            temp_path = tempfile.mkdtemp()
            
            mission_name = world
            if 'DEVBUILD' in mission_name:
                mission_name = mission_name.replace('DEVBUILD',VERSION)

            print('Creating training mission {}... ({}/{})'.format(color_string(world,'\033[92m',args.color), count+1, len(all_templates)))
            template_path = '{}/{}/{}'.format(templateDir, template_dir_name, world)
            try:
                content_list = fetch_objects(template_path)
                folder_list = content_list[0]
                file_list = content_list[1]
            except:
                sys.exit('No mission named \'{1}\' file was discoverd in the given path: ../{0}'.format(template_dir_name, world))

            os.chdir(template_path)

            for obj in folder_list:
                shutil.copytree(obj, '{}/{}'.format(temp_path, obj))
            for obj in file_list:
                shutil.copy2(obj, temp_path)

            # Unzip and install script package
            install_script_package(PACKAGE, temp_path, args.color)

            # get json data
            training_json = '{}/setup.json'.format(temp_path, world)
            with open(training_json) as json_file:
                training_json = json.load(json_file)

            # Setup mission file
            print('Setting up and adjusting training mission file...')
            setup_missions(temp_path, training_json, count, args.color)

            # Cleaning mission from readme and template files.
            clean_mission_folder(temp_path)

            # Building PBO
            build_pbo(temp_path, mission_name, args.color)

    build_archive('Mission_{}_v{}'.format(args.buildtype, VERSION), 'zip', outputDir, args.color)
    cleanup_output()

    print('Builds complete. ({} seconds)'.format(round(time.time() - startTime, 3)))

if __name__ == "__main__":
    sys.exit(main())

